pkg/insightstore/datapoint.go

```diff
@@ -0,0 +1,191 @@
+type DeployFrequency struct {
+	Timestamp   int64   `json:"timestamp"`
+	DeployCount float32 `json:"deploy_count"`
+}
+
+func (d DeployFrequency) GetTimestamp() int64 {
+	return d.Timestamp
+}
+
+func (d DeployFrequency) Value() float32 {
+	return d.DeployCount
+}
+
+// ChangeFailureRate represents a data point that shows the change failure rate metrics.
+type ChangeFailureRate struct {
+	Timestamp    int64   `json:"timestamp"`
+	Rate         float32 `json:"rate"`
+	SuccessCount int64   `json:"success_count"`
+	FailureCount int64   `json:"failure_count"`
+}
+
+func (c ChangeFailureRate) GetTimestamp() int64 {
+	return c.Timestamp
+}
+
+func (c ChangeFailureRate) Value() float32 {
+	return c.Rate
+}
+
+type DataPoint interface {
+	// Value gets data for model.InsightDataPoint
+	Value() float32
+	// Timestamp gets timestamp
+	GetTimestamp() int64
+}
+
+func Merge(dp1 DataPoint, dp2 DataPoint, kind model.InsightMetricsKind) (DataPoint, error) {
+	if dp1 == nil {
+		return dp2, nil
+	}
+	if dp2 == nil {
+		return dp1, nil
+	}
+
+	switch kind {
+	case model.InsightMetricsKind_DEPLOYMENT_FREQUENCY:
+		df1, ok := dp1.(DeployFrequency)
+		if !ok {
+			return nil, fmt.Errorf("cannot cast to DeployFrequency. DataPoint: %v", dp1)
+		}
+
+		df2, ok := dp2.(DeployFrequency)
+		if !ok {
+			return nil, fmt.Errorf("cannot cast to DeployFrequency. DataPoint: %v", dp1)
+		}
+
+		if df1.Timestamp != df2.Timestamp {
+			return nil, fmt.Errorf("mismatch timestamp. dp1: %d, dp2: %d", df1.Timestamp, df2.Timestamp)
+		}
+
+		df1.DeployCount += df2.DeployCount
+		return df1, nil
+	case model.InsightMetricsKind_CHANGE_FAILURE_RATE:
+		cfr1, ok := dp1.(ChangeFailureRate)
+		if !ok {
+			return nil, fmt.Errorf("cannot cast to ChangeFailureRate. DataPoint: %v", dp1)
+		}
+
+		cfr2, ok := dp2.(ChangeFailureRate)
+		if !ok {
+			return nil, fmt.Errorf("cannot cast to ChangeFailureRate. DataPoint: %v", dp2)
+		}
+
+		if cfr1.Timestamp != cfr2.Timestamp {
+			return nil, fmt.Errorf("mismatch timestamp. dp1: %d, dp2: %d", cfr1.Timestamp, cfr2.Timestamp)
+		}
+
+		cfr1.FailureCount += cfr2.FailureCount
+		cfr1.SuccessCount += cfr2.SuccessCount
+		cfr1.Rate = float32(cfr1.FailureCount) / float32(cfr1.FailureCount+cfr1.SuccessCount)
+		return cfr1, nil
+	default:
+		return nil, fmt.Errorf("invalid kind: %v", kind)
+	}
+}
+
+// convert types to list of DataPoint.
+func ToDataPoints(i interface{}) ([]DataPoint, error) {
+	switch dps := i.(type) {
+	case []DeployFrequency:
+		dataPoints := make([]DataPoint, len(dps))
+		for j, dp := range dps {
+			dataPoints[j] = dp
+		}
+		return dataPoints, nil
+	case []ChangeFailureRate:
+		dataPoints := make([]DataPoint, len(dps))
+		for j, dp := range dps {
+			dataPoints[j] = dp
+		}
+		return dataPoints, nil
+	default:
+		return nil, fmt.Errorf("cannot convert to DataPoints: %v", dps)
+	}
+}
+
+// AppendDataPoint append new data point to the end of list
+func AppendDataPoint(dp []DataPoint, point DataPoint) []DataPoint {
```

Comment:
Do we need this function? This is just doing exactly what `append` does.

Reply:
Indeed, I remove this func.